pkg/cdi/default-cache.go

```diff
@@ -0,0 +1,59 @@
+/*
+   Copyright © 2024 The CDI Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package cdi
+
+import (
+	"sync"
+
+	oci "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+var (
+	defaultCache   *Cache
+	getDefaultOnce sync.Once
+)
+
+// GetDefaultCache returns the default CDI cache instance.
+func GetDefaultCache() *Cache {
+	getDefaultOnce.Do(func() {
+		defaultCache = newCache()
+	})
+	return defaultCache
+}
+
+// Configure applies options to the default CDI cache. Updates and refreshes
+// the default cache if options have changed.
+func Configure(options ...Option) error {
+	return GetDefaultCache().Configure(options...)
+}
```

Comment:
I keep thinking about how we can get around the issue of `GetDefaultCache()` initialising a cache with the default (i.e. possibly WRONG) options here before reconfiguring it. This is a difference between what we have here and the existing registry-based implementation where the correct options are applied the first time.

Is there a way we can do this here too? Does pulling the `Once.Do` logic into a private function and has similar `new` logic to the existing `GetRegistry` function achieve this?

Reply:
What about something like:
```
diff --git a/pkg/cdi/default-cache.go b/pkg/cdi/default-cache.go
index 8975a0f..5c010d5 100644
--- a/pkg/cdi/default-cache.go
+++ b/pkg/cdi/default-cache.go
@@ -27,18 +27,29 @@ var (
 	getDefaultOnce sync.Once
 )
 
-// GetDefaultCache returns the default CDI cache instance.
-func GetDefaultCache() *Cache {
+func getOrCreate(options ...Option) (*Cache, bool) {
+	var created bool
 	getDefaultOnce.Do(func() {
-		defaultCache = newCache()
+		defaultCache = newCache(options...)
+		created = true
 	})
-	return defaultCache
+	return defaultCache, created
+}
+
+// GetDefaultCache returns the default CDI cache instance.
+func GetDefaultCache() *Cache {
+	cache, _ := getOrCreate()
+	return cache
 }
 
 // Configure applies options to the default CDI cache. Updates and refreshes
 // the default cache if options have changed.
 func Configure(options ...Option) error {
-	return GetDefaultCache().Configure(options...)
+	cache, created := getOrCreate(options...)
+	if len(options) == 0 || created {
+		return nil
+	}
+	return cache.Configure(options...)
 }
 
 // Refresh explicitly refreshes the default CDI cache instance.
```

Reply:
@elezar That's a valid point, too. I updated the implementation like that.